Lib/pathlib/_os.py

```diff
@@ -166,48 +166,78 @@ def copyfileobj(source_f, target_f):
         write_target(buf)
 
 
-def magic_open(path, mode='r', buffering=-1, encoding=None, errors=None,
-               newline=None):
+def _open_reader(obj):
+    cls = type(obj)
+    try:
+        return cls.__open_reader__(obj)
+    except AttributeError:
+        if hasattr(cls, '__open_reader__'):
+            raise
+    raise TypeError(f"{cls.__name__} can't be opened for reading")
```

Comment:
`try` should contain only the operation operation that can fail (`.`):
```suggestion
    try:
        open_reader = cls.__open_reader__
    except AttributeError:
        raise TypeError(f"{cls.__name__} can't be opened for reading")
    else:
        return open_reader(obj)
```